pkg/k8s/watchers/node.go

```diff
@@ -117,3 +135,104 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
 	}
 	return nodeInterface.(*v1.Node).DeepCopy(), nil
 }
+
+// ciliumNodeUpdater implements the subscriber.Node interface and is used
+// to keep CiliumNode objects in sync with the node ones.
+type ciliumNodeUpdater struct {
+	k8sWatcher         *K8sWatcher
+	localNode          nodediscovery.LocalNode
+	kvStoreNodeUpdater nodediscovery.KVStoreNodeUpdater
+}
+
+func NewCiliumNodeUpdater(k8sWatcher *K8sWatcher, kvStoreNodeUpdater nodediscovery.KVStoreNodeUpdater, localNode nodediscovery.LocalNode) *ciliumNodeUpdater {
+	return &ciliumNodeUpdater{
+		k8sWatcher:         k8sWatcher,
+		kvStoreNodeUpdater: kvStoreNodeUpdater,
+		localNode:          localNode,
+	}
+}
+
+func (u *ciliumNodeUpdater) OnAddNode(newNode *v1.Node, swg *lock.StoppableWaitGroup) error {
+	u.updateCiliumNode(u.kvStoreNodeUpdater, u.localNode, newNode)
+
+	return nil
+}
+
+func (u *ciliumNodeUpdater) OnUpdateNode(oldNode, newNode *v1.Node, swg *lock.StoppableWaitGroup) error {
+	u.updateCiliumNode(u.kvStoreNodeUpdater, u.localNode, newNode)
```

Comment:
Maybe I'm missing something, but do you not need to check that `newNode` actually refers to the local node? I don't think we want the local cilium-agent to update labels of a CiliumNode object of a remote node?

Reply:
Correct, the agent is supposed to update labels and addresses only of the CiliumNode object for its local node, and this is guaranteed by the fact that the `Node` watcher (which invokes this logic) is only subscribed to events for its own node:

https://github.com/cilium/cilium/blob/f2ecb1aa8ba2a86d49b6288dc25842f2eabcd453/pkg/k8s/watchers/node.go#L57-L62